services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/NetworkIPAvailabilityServiceImpl.java

```diff
@@ -0,0 +1,246 @@
+import com.futurewei.alcor.vpcmanager.service.NetworkIPAvailabilityService;
+import com.futurewei.alcor.vpcmanager.service.VpcDatabaseService;
+import com.futurewei.alcor.vpcmanager.utils.IPUtil;
+import com.futurewei.alcor.web.entity.subnet.AllocationPool;
+import com.futurewei.alcor.web.entity.subnet.SubnetEntity;
+import com.futurewei.alcor.web.entity.subnet.SubnetWebJson;
+import com.futurewei.alcor.web.entity.vpc.NetworkIPAvailabilityEntity;
+import com.futurewei.alcor.web.entity.vpc.SubnetIPAvailabilityEntity;
+import com.futurewei.alcor.web.entity.vpc.VpcEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Service
+public class NetworkIPAvailabilityServiceImpl implements NetworkIPAvailabilityService {
+
+    private Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    @Autowired
+    VpcDatabaseService vpcDatabaseService;
+
+    @Value("${microservices.subnet.service.url}")
+    private String subnetUrl;
+
+    private RestTemplate restTemplate = new RestTemplate();
+
+    /**
+     * get network ip availability by vpc id
+     * @param vpcid
+     * @return
+     * @throws ResourceNotFoundException
+     * @throws ResourcePersistenceException
+     */
+    @Override
+    public NetworkIPAvailabilityEntity getNetworkIPAvailability(String vpcid) throws ResourceNotFoundException, ResourcePersistenceException {
+
+        NetworkIPAvailabilityEntity networkIPAvailabilityEntity = new NetworkIPAvailabilityEntity();
+
+        VpcEntity vpcEntity = this.vpcDatabaseService.getByVpcId(vpcid);
+        if (vpcEntity == null) {
+            throw new ResourceNotFoundException("vpc can not be found" + vpcid);
+        }
+
+        networkIPAvailabilityEntity = setNetworkIPAvailability(vpcEntity);
+
+        return networkIPAvailabilityEntity;
+    }
+
+    /**
+     * list all network ip availabilities
+     * @param queryParams
+     * @return
+     * @throws CacheException
+     * @throws ResourceNotFoundException
+     */
+    @Override
+    public List<NetworkIPAvailabilityEntity> getNetworkIPAvailabilities(Map<String, Object[]> queryParams) throws CacheException, ResourceNotFoundException {
+
+        List<NetworkIPAvailabilityEntity> networkIPAvailabilityEntities = new ArrayList<NetworkIPAvailabilityEntity>();
+
+        Map<String, VpcEntity> vpcStates = this.vpcDatabaseService.getAllVpcs(queryParams);
+
+        // Filter vpcStates by request parameters
+//        if (vpcId != null) {
+//            vpcStates = vpcStates.entrySet().stream()
+//                    .filter(state -> vpcId.equalsIgnoreCase(state.getValue().getId()))
```

Comment:
Time to remove the commented block if we are ready :-)

Reply:
Sure